content/blog/2023-12-11-rvv-llvm-gisel/index.md

```diff
@@ -0,0 +1,177 @@
++++
+title = "Support LLVM GlobalISel for RISC-V Vector Extension on Partial ALU Instructions"
+[extra]
+bio = """
+  Jiahan Xie is a first-year CS MS student at Cornell whose interests lie in Compilers and Computer Architecture.
+
+  Michael Maitland was a student in Cornell CS 6120 Advanced Compilers and he's now working at SiFive.
+"""
+[[extra.authors]]
+name = "Jiahan Xie"
+[[extra.authors]]
+name = "Michael Maitland"
++++
+
+# Introduction
+
+The open [RISC-V instruction set architecture (ISA)](https://riscv.org/technical/specifications/) has an interesting extension, [the RISC-V "V" Vector Extension](https://github.com/riscv/riscv-v-spec/blob/master/v-spec.adoc). The unique part of RISC-V Vector Extension is that its vector registers have flexible widths, VLEN, which makes programming in RISC-V Vector Extension agnostic to the vector register sizes. This feature really distinguishes the RISC-V Vector Extension from the traditional SIMD extensions, such as [x86 SSE](https://en.wikipedia.org/wiki/Streaming_SIMD_Extensions) (with a fixed size 128-bit vector length)/[AVX](https://en.wikipedia.org/wiki/Advanced_Vector_Extensions) (256-bit), and [Arm NEON](https://developer.arm.com/Architectures/Neon) (128-bit). The increasing vector lengths can pose a challenge to the traditional SIMD extensions as they have to address compatability and support all existing fixed size vector lengths in their ISAs. On the contrary, with the vector lengths agnostic principles, binary code generated by RISC-V assembly is automatically portable between different CPUs.
```

Comment:
If a processor implements RISC-V V extension, the vector code is portable to other CPU that support the V extension. However, if a processor implements say the ARM Neon extension, then the NEON code is portable to other ARM processors that support the Neon extension.

The real power of vector length agnostic is when it comes to loops. Take the following example:

```
for (i = 0; i < N; i++)
  A[i] = B[i] + C[i]
```

On RISC-V, we can generate the following (pseudo) code
```
ph:
  i = 0
body:
  vsetvli (N-i) // set the VL as N-i, or VLMAX if N-i is too large for the hardware to process
  vadd.vv a_ptr, b_ptr, c_ptr
  i += VL // Increment I by the number of elements processed in this iteration
  branch_if_done exit, body // decide whether to exit loop
exit:
  ret
```

On a SIMD architecture, it is more complicated. You need to have two versions of the loop: the vector loop, and a scalar remainder loop. If N-i is smaller than the number of elements processed in the vector loop, then the scalar remainder loop needs to be executed. What is the best size for the number of elements to be processed in the vector loop? It depends on whether N will often be large or small. You may not know this ahead of time.